lib.rs

```diff
@@ -0,0 +1,511 @@
+
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+/// Simple AMM pool contract
+///
+/// This contract is based on Balancer multi asset LP design and all formulas are taken from the Balancer's whitepaper (https://balancer.fi/whitepaper.pdf)
+/// It has one pool with PSP22 tokens with equal weights
+///
+/// Swaps can be performed between all pairs in the pool whitelisted for trading
+/// Liquidity provisioning is limited to designated accounts only and works as deposits / withdrawals of arbitrary composition.
+#[ink::contract]
+mod amm_pool {
+    use ink::{contract_ref, prelude::{vec, vec::Vec}, storage::Mapping};
+    use psp22::{PSP22Error, PSP22};
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct SwapPair {
+        pub from: AccountId,
+        pub to: AccountId,
+    }
+
+    impl SwapPair {
+        pub fn new(from: AccountId, to: AccountId) -> Self {
+            Self { from, to }
+        }
+    }
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum AmmError {
+        PSP22(PSP22Error),
+        InsufficientAllowanceOf(AccountId),
+        Arithmethic,
+        WrongParameterValue,
+        TooMuchSlippage,
+        NotAdmin,
+        NotEnoughLiquidityOf(AccountId),
+        UnsupportedSwapPair(SwapPair),
+    }
+
+    impl From<PSP22Error> for AmmError {
+        fn from(e: PSP22Error) -> Self {
+            AmmError::PSP22(e)
+        }
+    }
+
+    #[ink(event)]
+    pub struct Withdrawn {
+        #[ink(topic)]
+        caller: AccountId,
+        #[ink(topic)]
+        token: AccountId,
+        amount: Balance,
+    }
+
+    #[ink(event)]
+    pub struct SwapPairAdded {
+        #[ink(topic)]
+        pair: SwapPair,
+    }
+
+    #[ink(event)]
+    pub struct SwapPairRemoved {
+        #[ink(topic)]
+        pair: SwapPair,
+    }
+
+    #[ink(event)]
+    pub struct Swapped {
+        caller: AccountId,
+        #[ink(topic)]
+        token_in: AccountId,
+        #[ink(topic)]
+        token_out: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(storage)]
+    pub struct AmmPool {
+        pub owner: AccountId,
+        pub swap_fee_percentage: u128,
```

Comment:
Where do we use `swap_fee_percentage`? I see that we can change it, but for some reason this is not taken into account in the `swap()` function

Reply:
It is referred to in the readme, it is to be implemented for a specific use case. The main point of the tutorial is showcasing the amm with multiple tokens, based on the Balancer model